### PR TITLE
feat(FormControl): support ruleDependencies

### PR DESCRIPTION
**Pull request by myNameIsDu**
before 

https://github.com/rsuite/rsuite/assets/53305259/0a4d07b7-ee99-4c52-84ca-08f22a91063c

after

https://github.com/rsuite/rsuite/assets/53305259/594a19e7-7271-44b1-82ab-54001add56d7

### CODE REVIEW

File: src/FormControl/test/FormControlSpec.tsx [hunk 1]
```diff
@@ -550,4 +550,35 @@ describe('FormControl', () => {
       });
     });
   });
+
+  describe('ruleDependencies', () => {
+    it('Should validate the field when the dependency changes', () => {
```
Comment (simonguo) [inline src/FormControl/test/FormControlSpec.tsx:555]: Form provides a way to cross-check dependencies between fields.
https://github.com/rsuite/schema-typed?tab=readme-ov-file#multi-field-cross-validation

If you want to add validation rules through `rule`, you can use `Schema.Types.StringType().when(...)`

https://github.com/rsuite/schema-typed?tab=readme-ov-file#whencondition-schemaspec-schemadeclarationdatatype-errormsgtype--type

File: src/Form/Form.tsx [hunk 2]
```diff
@@ -347,18 +350,10 @@ const Form: FormComponent = React.forwardRef((props: FormProps, ref: React.Ref<F
...
-    setFormError(formError);
-    onError?.(formError);
-    onCheck?.(formError);
-  });
-
-  const handleFieldSuccess = useEventCallback((name: string) => {
-    removeFieldError(name);
+  const handleErrorChange = useEventCallback((nextError: Record<string, ReactNode>) => {
+    setFormError(nextError);
+    onError?.(nextError);
```
Comment (simonguo) [inline src/Form/Form.tsx:355]: Check whether the Error object has changed, if not, do not call `onError`.

```tsx
if (!shallowEqual(realFormError, nextError)) {
  onError?.(nextError);
}
```

**Review by simonguo — CHANGES_REQUESTED** [review]

File: src/Form/Form.tsx (hunk 2 above)
Comment (myNameIsDu) [reply src/Form/Form.tsx]: done

**Review by simonguo — CHANGES_REQUESTED** [review]

File: docs/pages/components/form/en-US/index.md [hunk 3]
```diff
@@ -193,6 +193,7 @@ cleanErrorForField: (fieldName: keyof E, callback?: () => void) => void;
 | plaintext              | boolean                                               | Make the control plaintext.                                                                                                                                                   |
 | readOnly               | boolean                                               | Make the control readonly.                                                                                                                                                    |
 | rule                   | checkType                                             | Current field verification rule. If it conflicts with the `<Form>`'s `model`, it get higher priority, [example](/components/form-validation/#field-level-verification-rules). |
+| ruleDependencies       | string[]                                              | The dependent fields for the check.                                                                                                                                           |
```
[the remark this answers is not in the document]
Comment (myNameIsDu) [reply docs/pages/components/form/en-US/index.md]: done

File: docs/pages/components/form/zh-CN/index.md [hunk 4]
```diff
@@ -180,8 +180,10 @@ cleanErrorForField: (fieldName: keyof E, callback?: () => void) => void;
 | plaintext              | boolean                                               | 使控件为纯文本。                                                                                                            |
 | readOnly               | boolean                                               | 使控件为只读。                                                                                                              |
 | rule                   | checkType                                             | 当前字段的校验规则，如果与`<Form>`的`model`冲突，以当前为准，[示例](/zh/components/form-validation/#field-级别的校验规则)。 |
+| ruleDependencies       | string[]                                              | 校验依赖的字段。                                                                                                            |
```
[the remark this answers is not in the document]
Comment (myNameIsDu) [reply docs/pages/components/form/zh-CN/index.md]: done

File: src/FormControl/FormControl.tsx [hunk 5]
```diff
@@ -71,6 +73,9 @@ export interface FormControlProps<P = any, ValueType = any>
 
   /** Validation rule */
   rule?: CheckType<unknown, any>;
+
+  /** The dependent fields for the check */
```
[the remark this answers is not in the document]
Comment (myNameIsDu) [reply src/FormControl/FormControl.tsx]: done

**Pull request closed without merging**